SimTracker/TrackerHitAssociation/plugins/ClusterSLOnGPU.cu

```diff
@@ -0,0 +1,233 @@
+         hh.detInd_d[i], hh.charge_d[i], 
+         hh.xg_d[i],hh.yg_d[i],hh.zg_d[i],hh.rg_d[i],hh.iphi_d[i], 
+         tk1[2],tk1[3],sl.n1_d[i],
+         tk2[2],tk2[3],sl.n2_d[i]
+        );
+
+}
+
+
+
+namespace clusterSLOnGPU {
+
+ constexpr uint32_t invTK = 0; // std::numeric_limits<int32_t>::max();
+
+ void printCSVHeader() {
+      printf("HIT: %s %s %s %s %s %s %s %s %s %s %s %s %s %s %s\n", "ev", "ind",
+         "det", "charge",	
+         "xg","yg","zg","rg","iphi", 
+         "tkId","pt","n1","tkId2","pt2","n2" 
+        );
+     }
+
+
+  std::atomic<int> evId(0);
+  std::once_flag doneCSVHeader;
+
+  Kernel::Kernel(cuda::stream_t<>& stream, bool dump) : doDump(dump) {
+    if (doDump) std::call_once(doneCSVHeader,printCSVHeader);
+    alloc(stream);
+  }
+
+
+  void
+  Kernel::alloc(cuda::stream_t<>& stream) {
+   cudaCheck(cudaMalloc((void**) & slgpu.links_d,(ClusterSLGPU::MAX_DIGIS)*sizeof(std::array<uint32_t,4>)));
+
+   cudaCheck(cudaMalloc((void**) & slgpu.tkId_d,(ClusterSLGPU::MaxNumModules*256)*sizeof(uint32_t)));
+   cudaCheck(cudaMalloc((void**) & slgpu.tkId2_d,(ClusterSLGPU::MaxNumModules*256)*sizeof(uint32_t)));
+   cudaCheck(cudaMalloc((void**) & slgpu.n1_d,(ClusterSLGPU::MaxNumModules*256)*sizeof(uint32_t)));
+   cudaCheck(cudaMalloc((void**) & slgpu.n2_d,(ClusterSLGPU::MaxNumModules*256)*sizeof(uint32_t)));
+
+
+   cudaCheck(cudaMalloc((void**) & slgpu.me_d, sizeof(ClusterSLGPU)));
+   cudaCheck(cudaMemcpyAsync(slgpu.me_d, &slgpu, sizeof(ClusterSLGPU), cudaMemcpyDefault, stream.id()));
+   cudaCheck(cudaDeviceSynchronize());
+
+  }
+
+  void
+  Kernel::zero(cudaStream_t stream) {
+   cudaCheck(cudaMemsetAsync(slgpu.tkId_d,invTK,(ClusterSLGPU::MaxNumModules*256)*sizeof(uint32_t), stream));
+   cudaCheck(cudaMemsetAsync(slgpu.tkId2_d,invTK,(ClusterSLGPU::MaxNumModules*256)*sizeof(uint32_t), stream));
+   cudaCheck(cudaMemsetAsync(slgpu.n1_d,0,(ClusterSLGPU::MaxNumModules*256)*sizeof(uint32_t), stream));
+   cudaCheck(cudaMemsetAsync(slgpu.n2_d,0,(ClusterSLGPU::MaxNumModules*256)*sizeof(uint32_t), stream));
+  }
+
+
+  void 
+  Kernel::algo(DigisOnGPU const & dd, uint32_t ndigis, HitsOnCPU const & hh, uint32_t nhits, uint32_t n, cuda::stream_t<>& stream) {
+    
+    /*
+    size_t pfs = 16*1024*1024;
+    // cudaDeviceSetLimit(cudaLimitPrintfFifoSize,pfs);
+    cudaDeviceGetLimit(&pfs,cudaLimitPrintfFifoSize);
+    std::cout << "cudaLimitPrintfFifoSize " << pfs << std::endl;
+    */
+
+    zero(stream.id());
+
+    ClusterSLGPU const & sl = slgpu;
+
+    int ev = ++evId;
+    int threadsPerBlock = 256;
+
+    int blocks = (nhits + threadsPerBlock - 1) / threadsPerBlock;
+    verifyZero<<<blocks, threadsPerBlock, 0, stream.id()>>>(ev, dd.me_d, hh.gpu_d, nhits, sl.me_d);
+
+
+    blocks = (ndigis + threadsPerBlock - 1) / threadsPerBlock;
+
+    assert(sl.me_d);
+    simLink<<<blocks, threadsPerBlock, 0, stream.id()>>>(dd.me_d,ndigis, hh.gpu_d, sl.me_d,n);
+    cudaStreamSynchronize(stream.id());
```

Comment:
it is needed for the dump below in case of other printf (can go inside the if)

Reply:
Why? `dumpLink` below is launched asynchronously on the same CUDA stream, so I'd expect it to work without this synchronization.

Reply:
is printf that requires syncronization to dump the buffer to host.
otherwise it will overwrite the circular one on device.
at least this is what I understood (and observed)

Reply:
Ok, so you want to protect against any potential earlier printf? Then yes, please move to inside the if (with a comment explaining the need).